viewing_party/party.py

```diff
@@ -1,23 +1,162 @@
+        else:
+            return None
+    else:
+        return None
+            
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    for i in range(len(user_data["watchlist"])):
+        if title == user_data["watchlist"][i]["title"]: 
+            user_data["watched"].append(user_data["watchlist"][i])
+            user_data["watchlist"].pop(i)
+            
+            return user_data
+    return user_data
+
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    if len(user_data["watched"]) >= 1:
+        sum = 0
+        for i in range(len(user_data["watched"])):
+            sum += user_data["watched"][i]["rating"]
+        return sum / len(user_data["watched"])   
+    else:
+        return 0.0
 
-
+def get_most_watched_genre(user_data):
+    if len(user_data["watched"]) >= 1:
+        genre_list = []
+        for i in range(len(user_data["watched"])):
+            genre_list.append(user_data["watched"][i]["genre"])
+        
+        counter = 0
+        most_frequent_genre = genre_list[0]
+        for i in genre_list:
+            curr_frequency = genre_list.count(i)
+            if(curr_frequency> counter):
+                counter = curr_frequency
+                most_frequent_genre = i
+ 
+        return most_frequent_genre
+    else:
+        return None
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    my_movies_watched = []
+    for movie in user_data["watched"]:
+        my_movies_watched.append(movie)
+    
+    friend_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_movies.append(movie)
+
+    unique_movies = []
 
+    for movie in my_movies_watched:
+        if movie not in friend_movies and movie not in unique_movies:
+            unique_movies.append(movie)
         
+    return unique_movies
+
+def get_friends_unique_watched(user_data):
+    my_movies_watched = []
+    for movie in user_data["watched"]:
+        my_movies_watched.append(movie)
+    
+    friend_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_movies.append(movie)
+
+    unique_movies = []
+
+    for movie in friend_movies:
+        if movie not in my_movies_watched and movie not in unique_movies:
+            unique_movies.append(movie)
+        
+    return unique_movies
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    recommended_movies = []

```

Comment:
So minor style nitpick, Python has no requirement that all local variables be defined at the top of the function, so I usually wait until the last possible moment to define them. There are two benefits to this. 

First is a chance to avoid unnecessary memory allocation. It doesn't really apply in this case, but if you had checks that made an early return before you actually ended up needing `recommended_movies`, you can avoid the memory allocation for `recommended_movies`. Not a big deal for an empty list, but it could be with larger data structures or objects that open other resources such as network connections.

The second is more on the readability side. By defining `recommended_movies` right when it gets filled in, it more obviously ties `recommended_movies` with the `for` loop that fills it. In your current code, it's not too far away, but if you imagine you have more checks of the passed-in values or other code for other tasks, it becomes easier to forget what `recommended_movies` started as.